snippets/zoe/contracts/test-loan.js

```diff
@@ -0,0 +1,141 @@
+import { makeFakePriceAuthority } from '@agoric/zoe/tools/fakePriceAuthority';
+import { makeSubscriptionKit } from '@agoric/notifier';
+
+test('loan contract', async t => {
+  const zoe = makeZoe(makeFakeVatAdmin().admin);
+
+  const contractBundle = await bundleSource(
+    require.resolve('@agoric/zoe/src/contracts/loan'),
+  );
+  const installation = await E(zoe).install(contractBundle);
+
+  const collateralKit = makeIssuerKit('moola');
+  const loanKit = makeIssuerKit('simoleans');
+
+  // Create autoswap installation and instance
+  const autoswapBundle = await bundleSource(
+    require.resolve('@agoric/zoe/src/contracts/autoswap'),
+  );
+  const autoswapInstallation = await E(zoe).install(autoswapBundle);
+
+  const { instance: autoswapInstance } = await E(zoe).startInstance(
+    autoswapInstallation,
+    harden({ Central: collateralKit.issuer, Secondary: loanKit.issuer }),
+  );
+
+  const issuerKeywordRecord = harden({
+    Collateral: collateralKit.issuer,
+    Loan: loanKit.issuer,
+  });
+
+  const timer = buildManualTimer(console.log);
+
+  const priceAuthority = makeFakePriceAuthority({
+    mathIn: collateralKit.amountMath,
+    mathOut: loanKit.amountMath,
+    priceList: [4, 2],
+    timer,
+  });
+
+  const doAddCollateral = _ => {};
+  const allCollateralAmount = collateralKit.amountMath.make(1000);
+  const x = loanKit.amountMath.make(1500);
+
+  // #region customMarginCall
+  E(priceAuthority)
+    .quoteWhenLT(allCollateralAmount, x)
+    .then(priceQuote => doAddCollateral(priceQuote));
+  // #endregion customMarginCall
+
+  const { subscription: periodAsyncIterable } = makeSubscriptionKit();
+
+  // #region lend
+  const terms = {
+    mmr: 150,
+    autoswapInstance,
+    priceAuthority,
+    periodAsyncIterable,
+    interestRate: 5,
+  };
+
+  const { creatorInvitation: lendInvitation } = await E(zoe).startInstance(
+    installation,
+    issuerKeywordRecord,
+    terms,
+  );
+
+  const maxLoan = loanKit.amountMath.make(1000);
+
+  const proposal = harden({
+    give: { Loan: maxLoan },
+  });
+
+  const payments = harden({
+    Loan: loanKit.mint.mintPayment(maxLoan),
+  });
+
+  const lenderSeat = await E(zoe).offer(lendInvitation, proposal, payments);
+
+  const borrowInvitation = await E(lenderSeat).getOfferResult();
+  // #endregion lend
+
+  // #region borrow
+  const borrowerProposal = harden({
+    want: { Loan: maxLoan },
+    give: { Collateral: allCollateralAmount },
+  });
+
+  const borrowerPayments = {
+    Collateral: collateralKit.mint.mintPayment(allCollateralAmount),
+  };
+  const borrowSeat = await E(zoe).offer(
+    borrowInvitation,
+    borrowerProposal,
+    borrowerPayments,
+  );
+
+  const borrowFacet = E(borrowSeat).getOfferResult();
+  // #endregion borrow
+
+  const invitationIssuer = await E(zoe).getInvitationIssuer();
+
+  // #region closeLoanInvitation
+  const closeLoanInvitation = await E(borrowFacet).makeCloseLoanInvitation();
```

Comment:
I think the example would be cleaner without this `await`. It would mean that you have to wait for the promise later, but I think that's better done off-screen.

Reply:
Ok, but then it's a promise for an invitation. Is that easier to explain?

Reply:
I wouldn't say that it's easier to explain, but I think it's better to explain it than not.